File: longest-substring-without-repeating-characters/yd.py

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if s is None:
            return 0
        beg = end = maxLen = 0
        cmap = {}
        for end in range(len(s)):
            if s[end] in cmap and beg <= cmap[s[end]]:
                beg = cmap[s[end]] + 1
            else:
                maxLen = max(maxLen, end - beg + 1)
            cmap[s[end]] = end
        return maxLen

File: longest-substring-without-repeating-characters/yd.java
    public class Solution {
        public int lengthOfLongestSubstring(String s) {
            if (s == null) {
                return 0;
            }
            int max = 0;
            int[] map = new int[256];
            int beg = 0;
            int end = 0;
            while (end < s.length()) {
                char c = s.charAt(end);
                map[c]++;
                while (map[c] > 1) {
                    map[s.charAt(beg)]--;
                    beg++;
                }
                max = Math.max(max, end - beg + 1);
                end++;
            }
            return max;
        }
    }
